### ISSUE TITLE
Change `selector-pseudo-*-allowed-list` not to include vendor prefixes automatically

### ISSUE THREAD

**Issue by carlosjeurissen** [stylelint/stylelint#7542]
### What is the problem you're trying to solve?

Currently, when just `hover` is specified in the allowlist, `-webkit-hover` will be allowed as well even tho `:-webkit-hover` is invalid. This is pretty unexpected for pseudo classes / elements. Considering:
- most of them do not have vendor prefixes
- you can not join them in one selector as this could invalidate the whole selector
- the vendor-prefix is named differently (`:-moz-focusring`, `:-webkit-full-screen`)

### What solution would you like to see?

It would make more sense to explicitly add desired vendor prefixed pseudo elements/classes to the allowlist.

This can potentially be a breaking change. However:
- usage of allowlists in pseudo elements / classes seems rare
- rarely an equivalent vendor prefix ever existed (I could only found any-link, for which one would use :visited and :link as fallback)
- this can be specified in the stylelint changelog
- migration is fairly easy
- no breaking issues will occur, just a false positive

---

- pairs
  - [x] `selector-pseudo-class-allowed-list` (already support prefixed pseudo-classes)
  - [x] `selector-pseudo-class-disallowed-list`
  - [x] `selector-pseudo-element-allowed-list` (already support prefixed pseudo-elements)
  - [x] `selector-pseudo-element-disallowed-list`
  - [x] `declaration-property-value-allowed-list`
  - [x] `declaration-property-value-disallowed-list`
  - [x] `property-allowed-list`
  - [x] `property-disallowed-list`
  - [x] `function-allowed-list`
  - [x] `function-disallowed-list`
- consistency
  - `value-no-vendor-prefix`
    - [x] regex
    - [x] string
  - `property-no-vendor-prefix`
    - [x] regex
    - [x] string
  - `selector-no-vendor-prefix`
    - [x] regex
    - [x] string
- [x] `declaration-property-max-values`

**Comment by ybiquitous** [comment]
@carlosjeurissen Thanks for opening the discussion. The suggestion seems to make sense to me. Vendor prefixes have been gradually less used, and people can easily migrate their config to keep the current behavior if they want. E.g.

```js
{
  "selector-pseudo-class-allowed-list": [/^(-webkit-|-moz-)?hover$/]
}
```

But indeed, this can be a breaking change, even if it is a false positive. So, I think it's better to change in the next major release.

**Comment by ybiquitous** [comment]
Ref https://github.com/stylelint/stylelint/blob/c49f9b5a7b45cb35c58b5bc1cd669733896d0183/docs/about/semantic-versioning.md?plain=1#L29-L33

**Comment by carlosjeurissen** [comment]
@ybiquitous if we want to be extra careful we can indeed wait until a major release for this change to be implemented. In this case it could make sense to offer an option to the pseudo-allowed-list rules named "allowVendorPrefixes", which would be true by default until a major release.

So that:
```json
"selector-pseudo-class-allowed-list": [[
  "hover", "-moz-last-node"
], {
  "includeVendorPrefixed": false
}]
```

Would not allow ":-moz-hover" or ":last-node" but would allow ""-moz-last-node" and "hover"

**Comment by ybiquitous** [comment]
The `selector-pseudo-class-whitelist` rule (before `selector-pseudo-class-allowed-list`) was added in 2c4bfa9a06956a950b2bee69eedf90cb6f66d835 and published in <https://github.com/stylelint/stylelint/releases/tag/7.1.0>. That was 2016, 8 years ago. The rule initially supported vendor prefixes. Probably, I imagine including vendor prefixes by default was more important at that time.

https://github.com/stylelint/stylelint/blob/6a987fca600bfe0e894035cea186e02e4a1d69d5/src/rules/selector-pseudo-class-whitelist/index.js#L41

I also don't have a use case, but I'm still concerned about the compatibility. Even if adding an option like `includeVendorPrefixed`, the option would become useless in the next major version, right?

**Comment by Mouvedia** [comment]
If we are going the option way, I would prefer something like `strict: true` or `lax: false` because

> or ":last-node"

wouldn't be expected to not be permitted for

>  "includeVendorPrefixed": false

You can't change the statu quo default boolean value **before** a major release though.

**Comment by carlosjeurissen** [comment]
> or ":last-node" wouldn't be expected to not be permitted for

@Mouvedia This would be expected if the allowlist only contains "-moz-last-node".

@ybiquitous Fair. The option in a next version would indeed be pretty useless unless people like it as a feature. Yet again I do not see a use case for this. Lets be safe and wait for a next major version and skip having an option for this.

Let me know if I can provide a PR and what the process for a delayed PR would be.

**Comment by ybiquitous** [comment]
I added this issue to the list in #7396, preparing 17.0.0.

**Comment by Mouvedia** [comment]
related: #7624

**Comment by Mouvedia** [comment]
see also

> […and add option `noVendorIgnore`…](https://github.com/stylelint/stylelint/issues/1739)

> [`allowVendorPrefixes`](https://github.com/stylelint/stylelint/issues/7524#issuecomment-1956291147)

> [`### `checkPrefixed: true | false` (default: `false`)`](https://github.com/stylelint/stylelint/blob/7b766f48f970699de088438be6b8974ddd0dc156/src/rules/property-no-unknown/README.md?plain=1#L97)

i.e. the name chosen will have to be consistent for all affected rules

**Comment by Mouvedia** [comment]
This has been fixed by #7525.
Hence I am gonna repurpose this issue to cover the other rules that are still affected.

**Comment by Mouvedia** [comment]
The 2 rules listed in the title were fixed by #7525.
Hence I am repurposing this issue to cover the other rules that are still affected.
There are 3 categories:

- the `*-no-vendor-prefix` rules which should be consistent and be fixed all at once just before the next major
e.g. fix regex of `property-no-vendor-prefix` and  `selector-no-vendor-prefix`
- the `*-list` which needs to be fixed by pairs (if both are affected)
- `declaration-property-max-values` which can be fixed separately

For `*-no-vendor-prefix` rules that have an `ignore*` secondary option I would advise to follow what has been done with #7650.
i.e. not having to list all the prefixed (`-moz-`/`-webkit-`/`-ms-`/`-o-`/etc.) properties/selectors/values for `string` is actually useful if you can opts out using a regular expression

**Comment by Mouvedia** [comment]
There are 3 types of rules that are affected:

- the `*-no-vendor-prefix` rules which should be consistent and be fixed all at once just before the next major
e.g. fix regex of `property-no-vendor-prefix` and  `selector-no-vendor-prefix`
- the `*-list` which needs to be fixed by pairs (if both are affected)
- `declaration-property-max-values` which can be fixed separately

For `*-no-vendor-prefix` rules that have an `ignore*` secondary option I would advise to follow what has been done with #7650.
i.e. not having to list all the prefixed (`-moz-`/`-webkit-`/`-ms-`/`-o-`/etc.) properties/selectors/values for `string` is actually quite convenient for an UX standpoint if you can opt out using a regular expression
